single-line-diagram/single-line-diagram-core/src/main/java/com/powsybl/sld/SingleLineDiagram.java

```diff
@@ -174,138 +128,71 @@ public static void draw(Graph graph, Path svgFile, LayoutParameters layoutParame
-        drawSubstation(network, id, writerForSvg, metadataWriter, layoutParameters, componentLibrary,
-                new HorizontalSubstationLayoutFactory(), new SmartVoltageLevelLayoutFactory(network),
-                initProvider, styleProvider, prefixId);
-    }
+    public static void drawSubstation(Network network, String substationId, Writer writerForSvg, Writer metadataWriter, Param param) {
+        LayoutParameters layoutParameters = param.getLayoutParameters();
+        VoltageLevelLayoutFactory voltageLevelLayoutFactory = param.getVoltageLevelLayoutFactory();
+        SubstationLayoutFactory substationLayoutFactory = param.getSubstationLayoutFactory();
 
-    private static void drawSubstation(Network network, String substationId, Writer writerForSvg, Writer metadataWriter, LayoutParameters layoutParameters, ComponentLibrary componentLibrary,
-                                       SubstationLayoutFactory sLayoutFactory, VoltageLevelLayoutFactory vLayoutFactory,
-                                       DiagramLabelProvider initProvider, StyleProvider styleProvider, String prefixId) {
-        Objects.requireNonNull(sLayoutFactory);
-        Objects.requireNonNull(vLayoutFactory);
+        Objects.requireNonNull(substationLayoutFactory);
+        Objects.requireNonNull(voltageLevelLayoutFactory);
 
         SubstationGraph substationGraph = new NetworkGraphBuilder(network).buildSubstationGraph(substationId);
-        sLayoutFactory.create(substationGraph, vLayoutFactory).run(layoutParameters);
-        draw(substationGraph, writerForSvg, metadataWriter, layoutParameters, componentLibrary, initProvider, styleProvider, prefixId);
-    }
-
-    public static void draw(Graph graph, Writer writerForSvg, Writer metadataWriter, LayoutParameters layoutParameters, ComponentLibrary componentLibrary,
-                            DiagramLabelProvider initProvider, StyleProvider styleProvider, String prefixId) {
-        draw(graph, writerForSvg, metadataWriter, new DefaultSVGWriter(componentLibrary, layoutParameters),
-                initProvider, styleProvider, prefixId);
+        substationLayoutFactory.create(substationGraph, voltageLevelLayoutFactory).run(layoutParameters);
+        DefaultSVGWriter svgWriter = new DefaultSVGWriter(param.getComponentLibrary(), param.getLayoutParameters(), param.getSvgParameters());
+        draw(substationGraph, writerForSvg, metadataWriter, svgWriter, param.createLabelProvider(network), param.createStyleProvider(network), param.getSvgParameters().getPrefixId());
     }
 
-    public static void draw(Graph graph, Writer writerForSvg, Writer metadataWriter, DefaultSVGWriter svgWriter,
-                            DiagramLabelProvider initProvider, StyleProvider styleProvider, String prefixId) {
+    public static void draw(Graph graph, Writer writerForSvg, Writer metadataWriter, SVGWriter svgWriter, LabelProvider labelProvider, StyleProvider styleProvider, String prefixId) {
```

Comment:
I think we should hide the svgWriter and therefore remove this method. Besides, exposing the providers is against the new design.

This a utility class, hence we want to keep methods simple (simpler than they were!), an advanced user who wants to use a custom SvgWriter can still do it, without using this class.

Reply:
I did not wish to expose the providers but the LabelProvider and the StyleProvider cannot be retrieved using an SldParameters object only. I need a Network object too. That is why I had to "build" those providers earlier, when I could still have access to the Network object.